task to install multiple languages

pallinger opened the pull request:
I created a task to install multiple languages. It sets the language diretory path, downloads the [dataverse-language-packs](https://github.com/GlobalDataverseCommunityConsortium/dataverse-language-packs) from git (configurable repo and branch), and zips the property files for the selected languages, uploads it and configures the selected languages in the given dataverse installation.

The configuration is done by the _dataverse.language_ property (example and default provided in defaults/main.yml), which contains a _dataverse.language.enabled_ property, which is false by default, thus disabling this function for not-interested users.

The contained tasks are _mostly_ idempotent, and try to skip admin interface calls if there are no changes compared to the previous runs.

pallinger committed:
task to install multiple languages

pallinger commented on tasks/main.yml:
```diff
@@ -15,12 +15,7 @@
-   - postinstall
-   - s3
-   - gui
+   - always
```
This may have side effects, although I do not think so. On the other hand, removing rows of code is almost always a good idea. :)

donsizemore commented:
In testing this, I hit the following during the `upload language pack` step:

`"Status code was 404 and not [200]: HTTP Error 404: Not Found", "redirected": false, "server": "Payara Server 5.2021.5 #badassfish", "status": 404, "url": "http://localhost:8080/api/admin/datasetfield/loadpropertyfiles"`

```
  Exception while invoking class org.glassfish.ejb.startup.EjbDeployer load method
java.lang.RuntimeException: EJB Container initialization error
        at org.glassfish.ejb.startup.EjbApplication.loadContainers(EjbApplication.java:237)
        at org.glassfish.ejb.startup.EjbDeployer.load(EjbDeployer.java:286)
        at org.glassfish.ejb.startup.EjbDeployer.load(EjbDeployer.java:104)
        at org.glassfish.internal.data.ModuleInfo.load(ModuleInfo.java:219)
        at org.glassfish.internal.data.ApplicationInfo.load(ApplicationInfo.java:336)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.prepare(ApplicationLifecycle.java:570)
        at com.sun.enterprise.v3.server.ApplicationLoaderService.processApplication(ApplicationLoaderService.java:416)
        at com.sun.enterprise.v3.server.ApplicationLoaderService.postConstruct(ApplicationLoaderService.java:248)
        at org.jvnet.hk2.internal.ClazzCreator.postConstructMe(ClazzCreator.java:303)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:351)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:463)
        at org.glassfish.hk2.runlevel.internal.AsyncRunLevelContext.findOrCreate(AsyncRunLevelContext.java:281)
        at org.glassfish.hk2.runlevel.RunLevelContext.findOrCreate(RunLevelContext.java:65)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2102)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:93)
        at org.jvnet.hk2.internal.ServiceHandleImpl.getService(ServiceHandleImpl.java:67)
        at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.oneJob(CurrentTaskFuture.java:1213)
        at org.glassfish.hk2.runlevel.internal.CurrentTaskFuture$QueueRunner.run(CurrentTaskFuture.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ExceptionInInitializerError
        at edu.harvard.iq.dataverse.validation.PasswordValidatorServiceBean.<init>(PasswordValidatorServiceBean.java:102)
        at edu.harvard.iq.dataverse.validation.__EJB31_Generated__PasswordValidatorServiceBean__Intf____Bean__.<init>(Unknown Source)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.lang.Class.newInstance(Class.java:584)
        at com.sun.ejb.containers.BaseContainer.instantiateOptionalEJBLocalBusinessObjectImpl(BaseContainer.java:3800)
        at com.sun.ejb.containers.StatelessSessionContainer.initializeHome(StatelessSessionContainer.java:216)
        at com.sun.ejb.containers.StatelessContainerFactory.createContainer(StatelessContainerFactory.java:63)
        at org.glassfish.ejb.startup.EjbApplication.loadContainers(EjbApplication.java:225)
        ... 20 more
Caused by: java.util.MissingResourceException: Can't find bundle for base name Bundle, locale en
        at java.base/java.util.ResourceBundle.throwMissingResourceException(ResourceBundle.java:2055)
        at java.base/java.util.ResourceBundle.getBundleImpl(ResourceBundle.java:1689)
        at java.base/java.util.ResourceBundle.getBundleImpl(ResourceBundle.java:1593)
        at java.base/java.util.ResourceBundle.getBundle(ResourceBundle.java:1284)
        at edu.harvard.iq.dataverse.util.BundleUtil.getResourceBundle(BundleUtil.java:99)
        at edu.harvard.iq.dataverse.util.BundleUtil.getResourceBundle(BundleUtil.java:83)
        at edu.harvard.iq.dataverse.util.BundleUtil.getStringFromBundle(BundleUtil.java:30)
        at edu.harvard.iq.dataverse.util.BundleUtil.getStringFromBundle(BundleUtil.java:26)
        at edu.harvard.iq.dataverse.validation.GoodStrengthRule.<clinit>(GoodStrengthRule.java:18)
        ... 31 more
```

This is with `payara.language.enable` set to `true`. What else would I modify?

qqmyers commented:
I've seen this (today actually) when I set the language dir and had not yet populated that dir with Bundle. (I wasn't trying the api call you have, just trying to run things and this broke Dataverse/404 for all pages). Possibly a catch 22 for this API call since it is trying to write to the lang dir I think. A workaround could be to write a Bundle_en.properties file (or the default machine language for other locales) - it might even be OK to have it empty (if not copy the Bundle.properties file) since I think there is logic to default back to Bundles.properties if a given key isn't found.

pallinger commented:
Ok, I will try to run it with an empty bundle directory. If it also produces this error, then I will put an extra step to copy the default bundle file there.

pallinger committed:
Put a default bundle in the language directory after creation, to avoid a payara startup failure.

pallinger commented:
A managed to reproduce the bug and fixed it using [qqmyers](https://github.com/qqmyers)'s suggestion: put the default bundle in the new language directory after creating it.

qqmyers commented:
FYI: Hoping to fix this from the Dataverse side: https://github.com/IQSS/dataverse/pull/8434 - not yet merged.

pallinger commented:
> FYI: Hoping to fix this from the Dataverse side: [IQSS/dataverse#8434](https://github.com/IQSS/dataverse/pull/8434) - not yet merged.

While [4bd6142](https://github.com/GlobalDataverseCommunityConsortium/dataverse-ansible/pull/208/commits/4bd6142af516061e820b554e2eb3494e11da63ff) certainly seems a workaround for that bug, but it also does no real harm even if that bug is fixed. The only possible side-effect is that it wastes 200kB on the server if English is not configured.

pallinger committed:
use {{ dataverse.api.location }} instead of hardcoded url
use {{ dataverse.language.lang_directory }} instead of hardcoded path
added dataverse.language.lang_directory to defaults

pallinger committed:
fix syntax error by correct quoting

pallinger commented:
I just pushed a syntax fix, you can test it now, thank you.

Commit:
#208 update group_vars files to allow testing

Commit:
#208 correct solr 8.11.1 checksums

pallinger committed:
Merge remote-tracking branch 'community/develop' into languages

Commit:
#208 resolve merge conflicts

pallinger committed:
Merge pull request #1 from GlobalDataverseCommunityConsortium/languages

update group_vars file

donsizemore commented:
Thank you!
